File: tests/FSharpx.Tests/ValidationTests.fs
﻿module FSharpx.Tests.ValidationTests

open System
open FsUnitTyped
open Microsoft.FSharp.Core
open FSharpx.Collections
open FSharpx.CSharpTests
open FSharpx
open FSharpx.Functional
open FSharpx.Choice
#nowarn "FS0044"
open FSharpx.Validation
open FSharpx.Nullable
open NUnit.Framework

let validator pred error value =
    if pred value
        then Choice1Of2 value
        else Choice2Of2 (NonEmptyList.singleton error)

let (==) = LanguagePrimitives.PhysicalEquality
let inline (!=) a b = not (a == b)

let nonNull e = validator ((!=) null) e
let notEqual a = validator ((<>) a)

let validateAddressLines =
    validator 
        (fun (a: Address) -> a.Line1 != null || a.Line2 == null) 
        "Line1 is empty but Line2 is not"

let validateAddress (a: Address) = 
    returnM a
    <* nonNull "Post code can't be null" a.Postcode
    <* validateAddressLines a


let greaterThan o = validator ((<?) o)

let lengthNotEquals l = validator (fun (x: string) -> x.Length <> l) "Invalid length"

let validateOrder (o: Order) =
    let nameNotNull = nonNull "Product name can't be null" o.ProductName
    let positiveCost n = greaterThan (0m).n (sprintf "Cost for product '%s' can't be negative" n) o.Cost
    nameNotNull >>= positiveCost |> Choice.map (konst o)

(*    validation {
        let! name = nonNull "Product name can't be null" o.ProductName
        let! _ = greaterThan (0m).n (sprintf "Cost for product '%s' must be positive" name) o.Cost
        return o
    } *)
    

let validateOrders c = seqValidator validateOrder c
    
[<Test>]
let ValidateCustomer() = 
    let customer = 
        Customer(
            Surname = "foo",
            Address = Address(Postcode = "1424"),
            Orders = ResizeArray([
                                    Order(ProductName = "Foo", Cost = (5m).n)
                                    Order(ProductName = "Bar", Cost = (-1m).n)
                                    Order(ProductName = null , Cost = (-1m).n)
                     ]))
    let result = 
        returnM customer
        <* nonNull "Surname can't be null" customer.Surname
        <* notEqual "foo" "Surname can't be foo" customer.Surname
        <* validateAddress customer.Address
        <* validateOrders customer.Orders
    match result with
    | Success c -> failwithf "Valid customer: %A" c
    | Failure errors -> 
        printfn "Invalid customer. Errors:\n%A" errors
        errors.Length |> shouldEqual 3
        errors |> shouldContain "Cost for product 'Bar' can't be negative"
        errors |> shouldContain "Product name can't be null"
        errors |> shouldContain "Surname can't be foo"

[<Test>]
let ``using ap``() =
  let customer = Customer()
  let result = 
    returnM (konst2 customer)
    |> Validation.ap (nonNull "Surname can't be null" customer.Surname)
    |> Validation.ap (notEqual "foo" "Surname can't be foo" customer.Surname)
  match result with
  | Success c -> failwithf "Valid customer: %A" c
  | Failure errors -> 
      printfn "Invalid customer. Errors:\n%A" errors
      errors.Length |> shouldEqual 1
      errors |> shouldContain "Surname can't be null"

[<Test>]
let ``validation with sum monoid``() =
    let v = Validation.CustomValidation (Monoid.sum())
    // count the number of broken rules
    let intValidator x = Choice.mapSecond (konst 1) x
    let notEqual a = notEqual a "" >> intValidator
    let lengthNotEquals l = lengthNotEquals l >> intValidator
    let validateString x = 
        Choice.returnM x
        |> v.apl (notEqual "hello" x)
        |> v.apl (lengthNotEquals 5 x)
    match validateString "hello" with
    | Success c -> failwithf "Valid string: %s" c
    | Failure e -> Assert.AreEqual(2, e)

[<Test>]
let ``validation with unit monoid``() =
    // using the unit monoid to avoid the overhead of concatenating error lists.
    let v = Validation.CustomValidation Monoid.unit

    // convert validator errors to unit
    let unitValidator x = Choice.mapSecond ignore x
    let notEqual a = notEqual a "" >> unitValidator
    let lengthNotEquals l = lengthNotEquals l >> unitValidator
    let validateString x = 
        Choice.returnM x
        |> v.apl (notEqual "hello" x)
        |> v.apl (lengthNotEquals 5 x)
    match validateString "hello" with
    | Success c -> failwithf "Valid string: %s" c
    | Failure () -> ()

[<Test>]
let ``using sequenceIgnore``() =
    let vsError = [ Choice1Of2 "ok"; Choice2Of2 (NonEmptyList.singleton "err") ]
    let vsOk = [ Choice1Of2 "ok1"; Choice1Of2 "ok2" ]

    let vError = Validation.sequenceIgnore vsError
    match vError with
    | Choice2Of2 errors ->
        CollectionAssert.AreEqual(errors, [ "err" ])
    | _ ->
        failwith "Validation must not succeed if there are errors"

    let vOk = Validation.sequenceIgnore vsOk
    match vOk with
    | Choice1Of2 () -> ()
    | Choice2Of2 _ -> failwith "Validation failed on success values"

[<Test>]
let ``using mapMIgnore``() =
    let okAndErr = [ "ok"; "err" ]
    let oks = [ "ok1"; "ok2" ]

    let validate = validator ((<>) "err") "error!"

    let vError = Validation.mapMIgnore validate okAndErr
    let vOk = Validation.mapMIgnore validate oks

    match vError with
    | Choice2Of2 errors ->
        CollectionAssert.AreEqual(errors, [ "error!" ])
    | _ ->
        failwith "Validation must not succeed if there are errors"

    match vOk with
    | Choice1Of2 () -> ()
    | Choice2Of2 _ -> failwith "Validation failed on success values"

File: tests/FSharpx.CSharpTests/ValidationTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.FSharp.Collections;
using Microsoft.FSharp.Core;
using NUnit.Framework;
using Errors = FSharpx.Collections.NonEmptyList<string>;
using FSharpx.Collections;

namespace FSharpx.CSharpTests {
    [TestFixture]
    public class ValidationTests {
        int doSomething(int userID, int id) {
            // fetch some other entity, do "stuff"
            return userID + id;
        }

        void setError(string e) {
            Console.WriteLine("Error: {0}", e);
        }

        const string req_userID = "a123";
        const string req_otherID = "b999";

        [Test]
        public void Test1_Imperative() {

            int userID;
            var userID_ok = int.TryParse(req_userID, out userID);
            if (!userID_ok) {
                setError("Invalid User ID");
            } else {
                int id;
                var id_ok = int.TryParse(req_otherID, out id);
                if (!id_ok) {
                    setError("Invalid ID");
                } else {
                    Console.WriteLine(doSomething(userID, id));
                }
            }
        }

        [Test]
        public void Test1_option() {
            var userID = FSharpOption.ParseInt(req_userID);
            if (!userID.HasValue()) {
                setError("Invalid User ID");
            } else {
                var otherID = FSharpOption.ParseInt(req_otherID);
                if (!otherID.HasValue()) {
                    setError("Invalid ID");
                } else {
                    Console.WriteLine(doSomething(userID.Value, otherID.Value));
                }
            }
        }


        [Test]
        public void Test1_either() {

            var somethingOrError =
                from userID in FSharpOption.ParseInt(req_userID).ToFSharpChoice("Invalid User ID")
                from id in FSharpOption.ParseInt(req_otherID).ToFSharpChoice("Invalid ID")
                select doSomething(userID, id);

            somethingOrError.Match(Console.WriteLine, setError);

        }

        void setErrors(IEnumerable<string> errors) {
            foreach (var e in errors)
                Console.WriteLine("Error: {0}", e);
        }

        [Test]
        public void Test2_imperative() {
            var errors = new List<string>();

            int userID;
            var userID_ok = int.TryParse(req_userID, out userID);
            if (!userID_ok)
                errors.Add("Invalid user ID");

            int id;
            var id_ok = int.TryParse(req_otherID, out id);
            if (!id_ok)
                errors.Add("Invalid ID");

            if (errors.Count > 0)
                setErrors(errors);
            else
                Console.WriteLine(doSomething(userID, id));
        }

        [Test]
        public void Test2_either() {
            var userID = FSharpOption.ParseInt(req_userID)
                .ToFSharpChoice(NonEmptyList.Singleton("Invalid User ID"));
            var id = FSharpOption.ParseInt(req_otherID)
                .ToFSharpChoice(NonEmptyList.Singleton("Invalid ID"));

            var doSomethingFunc = L.F((int a, int b) => doSomething(a, b));
            var curriedDoSomething = doSomethingFunc.Curry();
            var result = curriedDoSomething.ReturnValidation()
                .ApValidation(userID)
                .ApValidation(id);

            //var result = L.F((int a, int b) => doSomething(a,b))
            //    .Curry().PureValidate()
            //    .ApV(userID)
            //    .ApV(id);

            result.Match(Console.WriteLine, setErrors);
        }

        [Test]
        public void Test2_either_LINQ() {
            var userID = FSharpOption.ParseInt(req_userID)
                .ToFSharpChoice(NonEmptyList.Singleton("Invalid User ID"));
            var id = FSharpOption.ParseInt(req_otherID)
                .ToFSharpChoice(NonEmptyList.Singleton("Invalid ID"));

            var result =
                from a in userID
                join b in id on 1 equals 1
                select doSomething(a, b);

            result.Match(Console.WriteLine, setErrors);
        }

        static FSharpChoice<T, Errors> NonNull<T>(T value, string err) where T: class {
            return FSharpChoice.Validator<T>(x => x != null, err)(value);
        }

        static FSharpChoice<T, Errors> NotEqual<T>(T value, T other, string err) {
            return FSharpChoice.Validator<T>(v => !Equals(v, other), err)(value);
        }

        static FSharpChoice<Address, Errors> ValidateAddressLines(Address a) {
            if (a.Line1 != null || a.Line2 == null)
                return FSharpChoice.Ok(a);
            return FSharpChoice.Error<Address>("Line1 is empty but Line2 is not");
        }

        /// <summary>
        /// Same as FSharpChoice.Validator, for demo purposes only
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="pred"></param>
        /// <param name="err"></param>
        /// <returns></returns>
        static Func<T, FSharpChoice<T, Errors>> Validator<T>(Predicate<T> pred, string err) {
            return x => {
                if (pred(x))
                    return FSharpChoice.Ok(x);
                return FSharpChoice.Error<T>(err);
            };
        }

        static Func<Address, FSharpChoice<Address, Errors>> ValidateAddressLines2 = 
            Validator<Address>(x => x.Line1 != null || x.Line2 == null, "Line1 is empty but Line2 is not");

        static FSharpChoice<Address, Errors> ValidateAddress(Address a) {
            return from x in NonNull(a.Postcode, "Post code can't be null")
                   join y in ValidateAddressLines(a) on 1 equals 1
                   select a;
        }

        static FSharpChoice<T, Errors> GreaterThan<T>(T value, T other, string err) where T: IComparable<T> {
            var valueNull = Equals(null, value);
            var otherNull = Equals(null, other);
            if (valueNull && otherNull || valueNull != otherNull || value.CompareTo(other) > 0)
                return FSharpChoice.Ok(value);
            return FSharpChoice.Error<T>(err);
        }

        static FSharpChoice<T?, Errors> GreaterThan<T>(T? value, T? other, string err) where T: struct, IComparable<T> {
            if (!value.HasValue && !other.HasValue || value.HasValue != other.HasValue || value.Value.CompareTo(other.Value) > 0)
                return FSharpChoice.Ok(value);
            return FSharpChoice.Error<T?>(err);
        }

        static FSharpChoice<Order, Errors> ValidateOrder(Order o) {
            return
                from name in NonNull(o.ProductName, "Product name can't be null")
                from cost in GreaterThan(o.Cost, 0, string.Format("Cost for product '{0}' must be positive", name))
                select o;
        }

        static FSharpChoice<FSharpList<Order>, Errors> ValidateOrders(IEnumerable<Order> orders) {
            var zero = ListModule.Empty<Order>().ReturnValidation();
            return orders
                .Select(ValidateOrder)
                .Aggregate(zero, (e, c) => from a in e
                                           join b in c on 1 equals 1
                                           select a.Cons(b));
        }

        [Test]
        public void Test3() {
            var customer = new Customer {
                Address = new Address {
                    Postcode = "1424",
                },
                Orders = new[] {
                    new Order {
                        ProductName = "Foo",
                        Cost = 5,
                    },
                    new Order {
                        ProductName = "Bar",
                        Cost = -1,
                    },
                    new Order {
                        ProductName = null,
                        Cost = -1,
                    },
                }
            };
            var result =
                from surname in NonNull(customer.Surname, "Surname can't be null")
                join surname2 in NotEqual(customer.Surname, "foo", "Surname can't be foo") on 1 equals 1
                join address in ValidateAddress(customer.Address) on 1 equals 1
                join orders in ValidateOrders(customer.Orders) on 1 equals 1
                select customer;
            result.Match(c => Assert.Fail("Validation should have failed"),
                         Console.WriteLine);
        }

        [Test]
        public void AddressWithoutLine1() {
            var a = new Address { Postcode = "1000" };
            ValidateAddress(a)
                .Match(_ => { },
                       err => Assert.Fail("Validation should not have failed with errors {0}", err));
        }

        [Test]
        public void AddressWithLine1AndNoLine2() {
            var a = new Address { Postcode = "1000", Line1 = "Fake Street" };
            ValidateAddress(a)
                .Match(_ => { },
                       err => Assert.Fail("Validation should not have failed with errors {0}", err));
        }

        [Test]
        public void AddressWithLine1AndLine2() {
            var a = new Address { Postcode = "1000", Line1 = "Fake Street", Line2 = "123" };
            ValidateAddress(a)
                .Match(_ => { },
                       err => Assert.Fail("Validation should not have failed with errors {0}", err));
        }

        [Test]
        public void AddressWithLine2AndNoLine1() {
            var a = new Address { Postcode = "1000", Line2 = "123" };
            ValidateAddress(a)
                .Match(_ => Assert.Fail("Validation should have failed"),
                       err => {
                           Assert.AreEqual(1, err.Length);
                           Assert.AreEqual("Line1 is empty but Line2 is not", err.Head);
                       });
        }
    }
}